extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -0,0 +1,348 @@
+			if (session.connection && session.connection.isMatch(connectionProfile)) {
+				return session;
+			}
+		}
+		return undefined;
+	}
+}
+
+export class Session {
+	private _root: RootNode;
+	constructor(private _connection: Connection, private sessionId?: string) {
+	}
+
+	public get uri(): string {
+		return this.sessionId || this._connection.uri;
+	}
+
+	public get connection(): Connection {
+		return this._connection;
+	}
+
+	public set root(node: RootNode) {
+		this._root = node;
+	}
+
+	public get root(): RootNode {
+		return this._root;
+	}
+}
+
+class RootNode extends TreeNode {
+	private children: TreeNode[];
+	constructor(private _session: Session, private context: TreeDataContext, private nodePath: string) {
+		super();
+	}
+
+	public get session(): Session {
+		return this._session;
+	}
+
+	public get nodePathValue(): string {
+		return this.nodePath;
+	}
+
+	public getChildren(refreshChildren: boolean): TreeNode[] | Promise<TreeNode[]> {
+		if (refreshChildren || !this.children) {
+			this.children = [];
+			let dataServicesNode = new DataServicesNode(this._session, this.context, this.nodePath);
+			dataServicesNode.parent = this;
+			this.children.push(dataServicesNode);
+		}
+		return this.children;
+	}
+
+	getTreeItem(): vscode.TreeItem | Promise<vscode.TreeItem> {
+		throw new Error('Not intended for use in a file explorer view.');
+	}
+
+	getNodeInfo(): sqlops.NodeInfo {
+		let nodeInfo: sqlops.NodeInfo = {
+			label: localize('rootLabel', 'Root'),
+			isLeaf: false,
+			errorMessage: undefined,
+			metadata: undefined,
+			nodePath: this.generateNodePath(),
+			nodeStatus: undefined,
+			nodeType: 'sqlCluster:root',
+			nodeSubType: undefined,
+			iconType: 'folder'
+		};
+		return nodeInfo;
+	}
+}
+
+class DataServicesNode extends TreeNode {
+	private children: TreeNode[];
+	constructor(private _session: Session, private context: TreeDataContext, private nodePath: string) {
+		super();
+	}
+
+	public get session(): Session {
+		return this._session;
+	}
+
+	public get nodePathValue(): string {
+		return this.nodePath;
+	}
+
+	public getChildren(refreshChildren: boolean): TreeNode[] | Promise<TreeNode[]> {
+		if (refreshChildren || !this.children) {
+			this.children = [];
+			let hdfsNode = new ConnectionNode(this.context, localize('hdfsFolder', 'HDFS'), this.createHdfsFileSource());
+			hdfsNode.parent = this;
+			this.children.push(hdfsNode);
+		}
+		return this.children;
+	}
+
+	private createHdfsFileSource(): IFileSource {
+		return this.session.connection.createHdfsFileSource();
+	}
+
+	getTreeItem(): vscode.TreeItem | Promise<vscode.TreeItem> {
+		throw new Error('Not intended for use in a file explorer view.');
+	}
+
+	getNodeInfo(): sqlops.NodeInfo {
+		let nodeInfo: sqlops.NodeInfo = {
+			label: localize('dataServicesLabel', 'Data Services'),
+			isLeaf: false,
+			errorMessage: undefined,
+			metadata: undefined,
+			nodePath: this.generateNodePath(),
+			nodeStatus: undefined,
+			nodeType: 'hadoop:root',
```

Comment:
Minor: I would prefer we rename this nodetype to be `'dataservices'` or similar, but we need to update any referencing package.json contributions too.